design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,195 @@
+# Solution Design - Inject Client Cert fix
+[//]: # "Change the title above from "Template" to your design's title"
+
+## Table of Contents
+[//]: # "You can use this tool to generate a TOC - https://ecotrust-canada.github.io/markdown-toc/"
+
+## Useful links
+[//]: # "Add links that may be useful for the reader"
+
+- [Existing PR with solution](https://github.com/cyberark/conjur/pull/1789)
+  - In this solution we wait for the timeout to get an answer and if we didn't
+    get one we create another WebSocket with an "ls" command to verify that the
+    cert is injected.
+- [Known issue in k8s exec](https://github.com/kubernetes/kubernetes/issues/89899)
+
+## Background
+[//]: # "Give relevant background for the designed feature. What is the motivation for this solution?"
+We have an issue with the "inject client cert" process where we don't always return a correct response to the client regarding the injection success. The injection is done via a WebSocket to Kubectl Exec where we use a "tar" command and the cert content in STDIN. There is [an issue](https://github.com/kubernetes/kubernetes/issues/89899) in k8s where messages that are sent via WebSockets to Kubectl Exec are left hanging without returning a response when using STDIN. 
+
+## Issue description
+[//]: # "Elaborate on the issue you are writing a solution for"
+The current behaviour is that we close the WebSocket connection after we send the injection request to kubectl. This is not correct as this way we do not wait for kubectl to send its response.
+
+This has 2 implications:
+1. If the injection request has succeeded we will return a 200 OK response to the 'inject_client_cert' request even if the cert has not finished to be written to the pod. This can lead to a race-condition where the client will try to read the cert before it is there.
+1. If the injection request has failed in the k8s API, we will not write the error to the Conjur logs, and we will falsely return a 200 OK response to the 'inject_client_cert' request
+
+## Solution
+[//]: # "Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breakdown of the feature, but just a recommendation to how the solution should be approached."
+
+We would like to inject the cert and get an indication that the cert was injected  or the error if it was not. Let's explore the different options that we have to inject the cert. 
```

Comment:
@orenbm I want to suggest another option.
Proposed solution: Run `sh` command instead of `tar`, and in the STDIN pass it a full script that will be executed.

The script will contain the whole content of the certificate, and just write it to the file.
It will also include any error handling, and the `exit` command closes the `sh` process which closes the websocket.
The script could be similar to:
```sh
#!/bin/sh

cd /etc/conjur/ssl

cat > conjur.pem <<EOF
#<SSL CLIENT CERTIFICATE>
EOF

exit
```
Benefits:
1. Short URL. Certificate is in the script. (avoid issues raised from option 2)
1. Deterministic operation. You know exactly when it is done, because the connection is closed once it reaches `exit`.
1. Custom error handling. You can get any error you wish for, handle it properly, and send it back.
1. Versatility. Script can be customized later if needed to handle more issues in this area.

Reply:
thanks @abrahamko ! adding this to the doc.

Reply:
This solution does require the container to have the `sh` executable.
I know that in secrets provider and authn k8s we have it.
Still, this requires that the customer must make sure the caller container has the `sh` executable.
@rafis3 , @sashaCher Can we require it from the customer? Is it risky?

Reply:
@abrahamko since the command is running on our sidecar container, then it could use the `sh` that is already available in that container. You don't need to expect the user to have it their app container.

Reply:
@rafis3 You're right, but this API is in the server, and so every client using it must be aware that this API requires `sh` executable available in its environment.
Our clients do, but other clients might not have it.

Reply:
That's true, but also true for `tar`. I don't know of anyone that chose to implement the authn client logic by themselves, so I'm relatively optimistic that this is not our biggest concern.

Reply:
i agree with the @rafis3 . the authn-client is a main part of the authentication process and the server side can rely on it.